File: src/anomaly/aggregator.ts
import { getPool } from "../db/index.js";
import type { ChainId } from "../types/index.js";

const FIVE_MIN_MS = 5 * 60 * 1000;
const TWENTY_FOUR_H_MS = 24 * 60 * 60 * 1000;
const FIVE_MIN_BUCKETS_IN_24H = (24 * 60) / 5; // 288

export interface ActiveToken {
  tokenAddress: string;
  swapCount: number;
}

/** 找过去 5min 有 swap 的活跃 token（白名单已经在入库时过滤过了） */
export async function findActiveTokens(chain: ChainId = "bsc"): Promise<ActiveToken[]> {
  const cutoff = Date.now() - FIVE_MIN_MS;
  const res = await getPool().query<{ token_address: string; swap_count: number }>(
    `SELECT token_address, SUM(swap_count)::int AS swap_count
     FROM token_1min_stats
     WHERE chain = $1 AND bucket_start >= $2
     GROUP BY token_address
     HAVING SUM(swap_count) > 0`,
    [chain, cutoff]
  );
  return res.rows.map((r) => ({
    tokenAddress: r.token_address,
    swapCount: r.swap_count,
  }));
}

export interface TokenWindowStats {
  vol5minUsd: number;
  fee5minUsd: number;
  vol24hUsd: number;
  vol24hAvg5minUsd: number;
  baselineCoverageMs: number; // 最早数据点到现在的距离，用来判断 baseline 是否可信
}

/** 一次性拿 token 的 5min 滚动 + 24h baseline 指标 */
export async function getTokenWindowStats(
  tokenAddress: string,
  chain: ChainId = "bsc"
): Promise<TokenWindowStats> {
  const now = Date.now();
  const fiveMinAgo = now - FIVE_MIN_MS;
  const twentyFourHAgo = now - TWENTY_FOUR_H_MS;

  const res = await getPool().query<{
    vol5min: string | null;
    fee5min: string | null;
    vol24h: string | null;
    earliest_bucket: string | null;
  }>(
    `SELECT
       COALESCE(SUM(total_volume_usd) FILTER (WHERE bucket_start >= $2), 0) AS vol5min,
       COALESCE(SUM(total_fees_usd)   FILTER (WHERE bucket_start >= $2), 0) AS fee5min,
       COALESCE(SUM(total_volume_usd) FILTER (WHERE bucket_start >= $3), 0) AS vol24h,
       MIN(bucket_start) FILTER (WHERE bucket_start >= $3) AS earliest_bucket
     FROM token_1min_stats
     WHERE token_address = $1 AND chain = $4`,
    [tokenAddress, fiveMinAgo, twentyFourHAgo, chain]
  );

  const row = res.rows[0];
  const vol5min = Number(row?.vol5min) || 0;
  const fee5min = Number(row?.fee5min) || 0;
  const vol24h = Number(row?.vol24h) || 0;
  const earliest = row?.earliest_bucket ? Number(row.earliest_bucket) : null;
  const baselineCoverageMs = earliest != null ? Math.max(0, now - earliest) : 0;

  // 用实际数据覆盖的 5min 窗口数算 baseline，避免冷启动期间被 288 系数稀释
  const effectiveWindows =
    baselineCoverageMs > 0
      ? Math.min(baselineCoverageMs / FIVE_MIN_MS, FIVE_MIN_BUCKETS_IN_24H)
      : 0;
  const vol24hAvg5min = effectiveWindows > 0 ? vol24h / effectiveWindows : 0;

  return {
    vol5minUsd: vol5min,
    fee5minUsd: fee5min,
    vol24hUsd: vol24h,
    vol24hAvg5minUsd: vol24hAvg5min,
    baselineCoverageMs,
  };
}

/** 找 token 涉及的池（V3 / Pancake-V3，V4 不含）+ 最近 30min 是否活跃 */
export async function findTokenActivePools(
  tokenAddress: string,
  chain: ChainId = "bsc",
  activeWindowMs: number = 30 * 60_000
): Promise<string[]> {
  const cutoff = Date.now() - activeWindowMs;
  const res = await getPool().query<{ pool_address: string }>(
    `SELECT DISTINCT p.address AS pool_address
     FROM pools p
     JOIN pool_1min_stats s
       ON s.pool_address = p.address AND s.chain = p.chain
     WHERE p.chain = $1
       AND (LOWER(p.token0) = LOWER($2) OR LOWER(p.token1) = LOWER($2))
       AND s.bucket_start >= $3`,
    [chain, tokenAddress, cutoff]
  );
  return res.rows.map((r) => r.pool_address);
}

/** 查最近 cooldown 内同 token+rule 是否已经触发过 */
export async function isRecentlyAlerted(
  tokenAddress: string,
  rule: string,
  cooldownMs: number
): Promise<boolean> {
  const cutoff = Date.now() - cooldownMs;
  const res = await getPool().query<{ exists: boolean }>(
    `SELECT EXISTS (
       SELECT 1 FROM anomaly_events
       WHERE token_address = $1 AND rule = $2 AND detected_at >= $3
     ) AS exists`,
    [tokenAddress.toLowerCase(), rule, cutoff]
  );
  return res.rows[0]?.exists ?? false;
}

File: src/anomaly/aggregator.rs
//! Detector 用的聚合查询：active tokens（有最近 swap）+ 5min/24h baseline + cooldown

use crate::db::get_pool;
use crate::types::CHAIN_BSC;
use anyhow::Result;

#[derive(Debug, Clone)]
pub struct TokenBaseline {
    pub token_address: Vec<u8>,    // BYTEA(20)
    pub symbol: Option<String>,
    pub vol_5min_usd: f64,
    pub vol_24h_avg_5min_usd: f64, // 24h volume / 288 (288 个 5min 桶)
    pub fee_5min_usd: f64,
    pub baseline_coverage_ms: i64,
}

/// 拿到最近 5min 内有 swap 活动的 token + 24h baseline
/// 新策略：symbol 来自 main_pools.coin（不再 JOIN binance_bsc_tokens）
pub async fn fetch_active_token_baselines(now_ms: i64) -> Result<Vec<TokenBaseline>> {
    let client = get_pool().get().await?;
    let win_5min_start = now_ms - 5 * 60_000;
    let win_24h_start = now_ms - 24 * 60 * 60_000;

    let rows = client.query(
        "WITH active_5min AS (
             SELECT token_address, SUM(total_volume_usd) AS vol5, SUM(total_fees_usd) AS fee5
             FROM token_1min_stats
             WHERE chain = $3 AND bucket_start >= $1
             GROUP BY token_address
             HAVING SUM(total_volume_usd) > 0
         ),
         baseline_24h AS (
             SELECT token_address,
                    SUM(total_volume_usd) AS vol24h,
                    MIN(bucket_start) AS earliest_bucket
             FROM token_1min_stats
             WHERE chain = $3 AND bucket_start >= $2
             GROUP BY token_address
         )
         SELECT a.token_address, b.vol24h, b.earliest_bucket, a.vol5, a.fee5,
                mp.coin
         FROM active_5min a
         LEFT JOIN baseline_24h b ON a.token_address = b.token_address
         LEFT JOIN main_pools mp ON mp.token_addr = a.token_address",
        &[&win_5min_start, &win_24h_start, &CHAIN_BSC],
    ).await?;

    let mut out = Vec::with_capacity(rows.len());
    for r in &rows {
        let token_addr_bytes: &[u8] = r.get(0);
        let vol_24h: Option<f64> = r.get(1);
        let earliest_bucket: Option<i64> = r.get(2);
        let vol_5min: f64 = r.get::<_, Option<f64>>(3).unwrap_or(0.0);
        let fee_5min: f64 = r.get::<_, Option<f64>>(4).unwrap_or(0.0);
        let symbol: Option<String> = r.get(5);

        let coverage_ms = match earliest_bucket {
            Some(eb) => now_ms - eb,
            None => 0,
        };
        let avg_5min = vol_24h.unwrap_or(0.0) / 288.0;

        out.push(TokenBaseline {
            token_address: token_addr_bytes.to_vec(),
            symbol,
            vol_5min_usd: vol_5min,
            vol_24h_avg_5min_usd: avg_5min,
            fee_5min_usd: fee_5min,
            baseline_coverage_ms: coverage_ms,
        });
    }
    Ok(out)
}

/// 检查 token 在 cooldown 内是否触发过（避免短时重复告警）
pub async fn is_in_cooldown(token: &[u8], rule: i16, since_ms: i64) -> Result<bool> {
    let client = get_pool().get().await?;
    let row = client.query_opt(
        "SELECT 1 FROM anomaly_events WHERE token_address = $1 AND rule = $2 AND detected_at >= $3 LIMIT 1",
        &[&token, &rule, &since_ms],
    ).await?;
    Ok(row.is_some())
}

pub async fn record_anomaly(
    token_address: &[u8],
    symbol: Option<&str>,
    rule: i16,
    metrics: serde_json::Value,
    detected_at: i64,
) -> Result<()> {
    let client = get_pool().get().await?;
    client.execute(
        "INSERT INTO anomaly_events (token_address, symbol, rule, metrics, detected_at) VALUES ($1, $2, $3, $4, $5)",
        &[&token_address, &symbol, &rule, &metrics, &detected_at],
    ).await?;
    Ok(())
}
